static/js/stream_list.js

```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+
+        var next_sibling = next_sibing_in_dir(current_sel, dir_up);
+
+        if (highlighting_class === 'highlighted_stream'
+            && next_sibling.is('hr.stream-split')) {
+            // Only for the left sidebar
+            // Skip separator
+            next_sibling = next_sibing_in_dir(next_sibling, dir_up);
+        }
 
-        var top_stream_id = $('#stream_filters li.narrow-filter').first().data('stream-id');
-        // undefined if there are no results
-        if (top_stream_id !== undefined) {
-            var top_stream = stream_data.get_sub_by_id(top_stream_id);
-            if (overlays.is_active()) {
-                ui_util.change_tab_to('#home');
+        if (!next_sibling.is('li.narrow-filter')) {
+            // At the every bottom or top
+            var all_streams = $(all_streams_selector);
+            if (dir_up) {
+                // top -> start at the bottom
+                next_sibling = all_streams.last();
+            } else {
+                // bottom -> start at the top
+                next_sibling = all_streams.first();
             }
-            exports.clear_and_hide_search();
-            narrow.by('stream', top_stream.name,
-                      {select_first_unread: true, trigger: 'sidebar enter key'});
-            e.preventDefault();
-            e.stopPropagation();
         }
+
+        // Classes must be explicitly named
+        if (highlighting_class === 'highlighted_stream') {
+            current_sel.removeClass('highlighted_stream');
+            next_sibling.addClass('highlighted_stream');
+
+        } else if (highlighting_class === 'highlighted_user') {
+            current_sel.removeClass('highlighted_user');
+            next_sibling.addClass('highlighted_user');
+        }
+
+        exports.scroll_element_into_container(next_sibling, scroll_container);
+    }
+}
+
+
+exports.keydown_filter = function (e, all_streams_selector,
+ scroll_container, highlighting_class, enter_press_function) {
+    // Function for left and right sidebar
+    // Could be placed somewhere else but ui.js is already very full
+
+    // Catch <enter> and <up-arrow>, <down-arrow> key presses
+    var handled = false;
+
+    switch (e.keyCode) {
+        case 13: {
+            // Enter key was pressed
+            enter_press_function();
+            handled = true;
+            break;
+        }
+        case 38: {
+            // Up-arrow key was pressed
+            keydown_arrow_key(true, all_streams_selector,
+                             scroll_container, highlighting_class);
+            handled = true;
+            break;
+        }
+        case 40: {
+            // Down-arrow key was pressed
+            keydown_arrow_key(false, all_streams_selector,
+                             scroll_container, highlighting_class);
+            handled = true;
+            break;
+        }
+    }
+
+    if (handled) {
+        // Not handle key event (do not move cursor)
```

Comment:
This comment would be better as something like "Since we already handled the key event above, suppress the browser handling it."; as written it can be read as the opposite of its intent ;)